### PR TITLE
feature(js): Adds a fixed Ajax spinner module

### PR DESCRIPTION
**Pull request by mrclay**
I think I based this on a Google site. A fixed spinner seems like a better general solution than figuring out where to put them within a page.
- [x] Do not use inline js
- [x] Document how to use the spinner
- [x] Test

### CODE REVIEW

File: mod/aalborg_theme/views/default/css/elements/misc.php [hunk 1]
```diff
@@ -62,3 +62,22 @@
...
 	padding: 0;
 }
+
+/* ***************************************
+	GLOBAL AJAX SPINNER
+*************************************** */
+
+.elgg-wait-cursor,
+.elgg-wait-cursor * {
+	cursor:wait !important
```
Comment (juho-jaakkola) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: Why does this need to use ```!important```? Usually the need to use it means that there's something wrong with the CSS elsewhere.

File: views/default/css/elements/misc.php [hunk 2]
```diff
@@ -59,3 +59,5 @@
 	display: none;
 	padding: 0;
 }
+
+<?php echo elgg_view('css/elements/spinner.css') ?>
```
Comment (juho-jaakkola) [inline views/default/css/elements/misc.php]: Why is this not called in ```views/default/css/elgg```?

File: mod/aalborg_theme/views/default/css/elements/misc.php [hunk 3]
```diff
@@ -62,3 +62,22 @@
...
+.elgg-spinner {
+	position:fixed;
+	top:0;
+	left:50%;
+	z-index:10000;
+	padding:10px;
+	border-radius:0 0 10px 10px;
+	border:1px solid #CCC;
+	border-top:0
+}
```
Comment (juho-jaakkola) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: Why is this added to Aalborg directly, but in core through ```css/elements/misc```?

File: mod/aalborg_theme/views/default/css/elements/misc.php (hunk 1 above)
Comment (mrclay) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: This was mostly pulled as is from a site. The important probably isn't necessary.

File: views/default/css/elements/misc.php (hunk 2 above)
Comment (mrclay) [inline views/default/css/elements/misc.php]: @ewinslow had brought up the idea of separating CSS into more files. For BC those would be pulled in from the original css/elgg/elements/*.php files, so this was for consistency with that plan I guess

File: mod/aalborg_theme/views/default/css/elements/misc.php (hunk 3 above)
Comment (mrclay) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: Mistake :)

File: mod/aalborg_theme/views/default/css/elements/misc.php (hunk 1 above)
Comment (juho-jaakkola) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: Might be best to check the other properties also and remove anything that's not needed by Elgg.

File: views/default/css/elements/misc.php (hunk 2 above)
Comment (juho-jaakkola) [inline views/default/css/elements/misc.php]: Ok, I see. This file wouldn't introduce anything non-BC, though?

**Comment by mrclay** [comment]
Before I finish this are there any objections to adding something like this? On the site I pulled it from we even use it for some non-ajaxed form submissions as it makes the UX seem a bit more responsive before the full page redraw begins.

File: mod/aalborg_theme/views/default/css/elements/misc.php (hunk 1 above)
Comment (mrclay) [inline mod/aalborg_theme/views/default/css/elements/misc.php]: Will do. Style doesn't match either.

File: views/default/css/elements/misc.php (hunk 2 above)
Comment (mrclay) [inline views/default/css/elements/misc.php]: I'm not sure what you're asking.

Comment (juho-jaakkola) [inline views/default/css/elements/misc.php]: I mean we could include this also in ```views/default/css/elgg```.

File: views/default/js/elgg/spinner.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+			num_running += 1;
+		},
+
+		/**
+		 * Remove an instance of the spinner. The indicator is displayed until all instances are removed.
+		 *
+		 * @param {jQuery} $wait       $-wrapped element(s) from which to remove the elgg-wait-cursor class
+		 * @param {Object} fadeOptions Options for the $.fadeOut method
+		 */
+		remove: function ($wait, fadeOptions) {
```
Comment (mrclay) [inline views/default/js/elgg/spinner.js]: I could get rid of remove() if I returned a Promise ($.Deferred) that the user would manually resolve().

File: engine/lib/views.php [hunk 5]
```diff
@@ -1604,6 +1604,8 @@ function elgg_views_boot() {
 	$elgg_js_url = elgg_get_simplecache_url('js', 'elgg');
 	elgg_register_js('elgg', $elgg_js_url, 'head');
 
+	elgg_extend_view('js/elgg', 'js/elgg/spinner.js', 501);
```
Comment (ewinslow) [inline engine/lib/views.php]: Prefer not to extend this view any more if possible. I know it creates more http requests but this is not something that should block page rendering anyways so I think that's OK.

File: views/default/js/elgg/spinner.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
+define('elgg/spinner', function (require) {
+	var $ = require('jquery'),
+		elgg = require('elgg'),
+
+		num_running = 0,
+		$spinner;
+
+	var exports = {
```
Comment (ewinslow) [inline views/default/js/elgg/spinner.js]: Add,remove,remove all? Smells like a collection api! Not sure what to do about that right now, just sort of interesting to me.

File: views/default/js/elgg/spinner.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
+define('elgg/spinner', function (require) {
```
Comment (ewinslow) [inline views/default/js/elgg/spinner.js]: Make this anonymous.

File: views/default/js/elgg/spinner.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		},
+
+		/**
+		 * Remove all instances of the spinner and remove the indicator.
+		 *
+		 * @param {Object} fadeOptions Options for the $.fadeOut method
+		 */
+		removeAll: function (fadeOptions) {
+			if (num_running) {
+				num_running = 1;
```
Comment (ewinslow) [inline views/default/js/elgg/spinner.js]: Is this supposed to be 0?

File: views/default/js/elgg/spinner.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+		add: function ($wait) {
+			if ($wait instanceof $) {
+				$wait.addClass('elgg-wait-cursor');
+			}
+
+			if (!num_running) {
+				if ($spinner) {
+					$spinner.show();
+				} else {
+					$spinner = $('<div class="elgg-spinner elgg-ajax-loader"></div>');
```
Comment (ewinslow) [inline views/default/js/elgg/spinner.js]: So is spinner a type of ajax loader?

File: views/default/js/elgg/spinner.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		 *
+		 * @param {jQuery} $wait       $-wrapped element(s) from which to remove the elgg-wait-cursor class
+		 * @param {Object} fadeOptions Options for the $.fadeOut method
+		 */
+		remove: function ($wait, fadeOptions) {
+			if (!num_running) {
+				return;
+			}
+
+			if ($wait instanceof $) {
```
Comment (ewinslow) [inline views/default/js/elgg/spinner.js]: Can we just assert this as a precondition and throw if not met? I prefer not to add branches to our  code if possible.

**Comment by ewinslow** [comment]
it would also be awesome if you could write a test for this.

Right now we're basically exporting a global object (global state). Have you considered exporting a class instead that gets instantiated elsewhere?

**Comment by mrclay** [comment]
Converted to constructor and documented

File: js/lib/ajax.js [hunk 11]
```diff
@@ -14,7 +14,7 @@ elgg.provide('elgg.ajax');
  *
  * @param {string} url Optionally specify the url as the first argument
  * @param {Object} options Optional. {@link jQuery#ajax}
- * @return {XmlHttpRequest}
+ * @return {jqXHR}
```
Comment (mrclay) [inline js/lib/ajax.js]: You think this needs a separate PR?

**Comment by mrclay** [comment]
Added some tests!

File: js/tests/ElggSpinnerTest.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+				}
+			});
+
+			spinner.start();
+			spinner.stop();
+		});
+
+		it("Indicator stays until all instances stopped", function() {
+			var s1 = new Spinner(),
+				s2 = new Spinner({
```
Comment (ewinslow) [inline js/tests/ElggSpinnerTest.js]: Tabs are making this not align well. Spaces for alignment?

File: js/lib/ajax.js (hunk 11 above)
Comment (ewinslow) [inline js/lib/ajax.js]: Separate commit maybe?

File: docs/guides/javascript.rst [hunk 13]
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
       }
    });
 
+Module elgg/Spinner
```
Comment (ewinslow) [inline docs/guides/javascript.rst]: Code ticks?

File: docs/guides/javascript.rst [hunk 14]
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
...
+             // ...
+          },
+          complete: function () {
+             spinner.stop();
+          }
+      });
+   });
+
+**Note:** The indicator is only removed once all Spinner instances are stopped.
+
```
Comment (ewinslow) [inline docs/guides/javascript.rst]: Thank you for the docs!

File: js/tests/ElggSpinnerTest.js (hunk 12 above)
Comment (ewinslow) [inline js/tests/ElggSpinnerTest.js]: Alternatively, just declare a new var on each line and let the minifier compress them (my preference).

**Comment by mrclay** [comment]
All should be taken care on this.

File: docs/guides/javascript.rst (hunk 14 above)
Comment (hypeJunction) [inline docs/guides/javascript.rst]: In my code, I prefer using beforeSend parameter to display spinners. It seems more organized this way.

File: docs/guides/javascript.rst [hunk 15]
```diff
@@ -339,6 +339,30 @@ Calls an Elgg action with the data passed. This handles outputting of system mes
       }
    });
 
+Module ``elgg/Spinner``
+~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (ewinslow) [inline docs/guides/javascript.rst]: Seems like sphinx doesn't like this header notation. It may work to switch to all dashes

File: mod/developers/views/default/js/elgg/dev.js [hunk 16]
```diff
@@ -2,10 +2,11 @@
...
  */
 define(function(require) {
-	var $ = require('jquery'); require('jquery.jstree');
+	var $ = require('jquery');
 
-	var setupInspectTree = function() {
-		
+	require('jquery.jstree');
+
+	function setupInspectTree() {
```
Comment (ewinslow) [inline mod/developers/views/default/js/elgg/dev.js]: This file can be or extracted to a separate commit/PR. It's cleanup that isn't relevant to this change.

File: views/default/css/admin.php [hunk 17]
```diff
@@ -1595,6 +1595,8 @@
 	margin: 0 0 1em 2em;
 }
 
+<?php echo elgg_view('css/elements/spinner.css') ?>
```
Comment (ewinslow) [inline views/default/css/admin.php]: I don't think this CSS has the right view name.

**Comment by ewinslow** [comment]
So I can't help but think that all the files for this component should be much more localized. It is a pretty standalone component but you literally had to edit 10+ files in very distant places to get it integrated.

Here's my overview of the pieces of this PR with the files that were modified.

* implementation
  * views/default/js/elgg/spinner.js (logic)
  * views/default/css/elements/misc/spinner.css (style)
* unit tests (js/tests/ElggSpinnerTest.js)
* developer docs (docs/guides/javascript.rst)
* demo
  * mod/developers/start.php
  * mod/developers/views/default/js/elgg/dev/theme_sandbox.js
  * mod/developers/views/default/theme_sandbox/javascript.php
  * mod/developers/views/default/theme_sandbox/javascript/spinner.php
* integration
  * views/default/css/admin.php
  * views/default/css/elgg.php

It feels like there is hardly any rhyme or reason to which files need editing. Shouldn't things be as simple as:

* Spinner.css -- styles
* Spinner.js -- implementation
* Spinner.rst -- dev docs
* SpinnerDemo.html -- demo templates
* SpinnerDemo.js -- demo scripts
* SpinnerTest.js -- unit tests

pick a folder, plop these all in, done.
??

File: views/default/css/admin.php [hunk 18]
```diff
@@ -1595,6 +1595,8 @@
 	margin: 0 0 1em 2em;
 }
 
+<?php echo elgg_view('css/elements/misc/spinner.css') ?>
```
Comment (mrclay) [inline views/default/css/admin.php]: @ewinslow fixed link

File: docs/guides/javascript.rst (hunk 14 above)
Comment (mrclay) [inline docs/guides/javascript.rst]: I changed it to use beforeSend, and switched this to reference the instance methods directly (also added a test to verify we can call these without `this` context)

**Comment by mrclay** [comment]
@ewinslow Yes, JS components having a single location would be great. We might be able to lean on grunt for this, but it would add a build step for regular devs.

**Comment by mrclay** [comment]
Since fadeOut is always async I found the JS tests weren't really working and were not run in isolation as I'd expected, so I rewrote them using [Jasmine's blocking functions](http://jasmine.github.io/1.3/introduction.html#section-Asynchronous_Support) and added a private module function to reset global state.

**Comment by mrclay** [comment]
I no longer think this API is right for core. I'd prefer just start() and stop() methods on a singleton module. No $wait or fade options. Fade using CSS animation.

**Comment by ewinslow** [comment]
Did you want to update and submit, or just close this PR?

**Comment by mrclay** [comment]
I'll make another PR.

**Comment by mrclay** [comment]
Replaced by #7741
